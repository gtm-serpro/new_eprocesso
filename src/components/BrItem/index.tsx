import { BRItem } from "@govbr-ds/core";
import classNames from "classnames";
import React, { memo, useCallback, useEffect, useRef, useState } from "react";
import CustomTag from "../../common/CustomTag";
import useCommonProperties from "../../hooks/useCommonProperties";
import { useMtProps } from "../../hooks/useMtProps";
import { useSpreadProps } from "../../hooks/useSpreadProps";
import type IMtProps from "../../types/IMtProps";
import BrDivider from "../BrDivider";
import BrList from "../BrList";

export interface BrItemProps extends React.HTMLAttributes<HTMLElement>, IMtProps {
  /** Se o item tem um highlight ao passar o mouse em cima. */
  highlighted?: boolean;
  /** Se o item tem um divider */
  divider?: boolean;
  /** Se o item está desabilitado. */
  disabled?: boolean;
  /** Target do item, referenciando o ID de uma lista, caso este item sirva para abrir/fechar a lista. */
  target?: string;
  /** Se abre/fecha. */
  collapsable?: boolean;
  /** Link do item */
  link?: string;

  /** Sub-lista de itens associados a este item. */
  subItems?: React.ReactNode;

  onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;

  /** Se é inicialmente expandido. */
  initiallyExpanded?: boolean;
}

export interface ItemRef extends HTMLDivElement {
  element: HTMLElement;
}

const BrItem = React.forwardRef<ItemRef, BrItemProps>(
  (
    {
      className,
      children,
      highlighted,
      divider = false,
      role = "listItem",
      disabled = false,
      target,
      collapsable = false,
      initiallyExpanded = false,
      link,
      subItems,
      onClick,
      ...props
    },
    ref
  ) => {
    const mtProps = useMtProps(props);
    const spreadProps = useSpreadProps(props);
    const refDiv = useRef<HTMLDivElement>(null);
    const refElemento = useRef(null);

    const [expanded, setExpanded] = useState<boolean>(initiallyExpanded);

    useCommonProperties<ItemRef>(ref, refDiv, {
      get element() {
        return refDiv.current;
      },
    });

    useEffect(() => {
      if (refDiv.current && !refElemento.current) {
        refElemento.current = new BRItem("br-item", refDiv.current);
      }
    }, []);

    const handleOnClick = useCallback(
      (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
        onClick?.(event);
        setExpanded((oldExpanded) => !oldExpanded);
      },
      [onClick]
    );

    return (
      <>
        <CustomTag
          ref={refDiv}
          tagName={link ? "a" : "div"}
          onClick={handleOnClick}
          className={classNames("br-item", highlighted && "highlighted", className, ...mtProps)}
          {...(role && { role: role })}
          {...(disabled && { disabled: true })}
          {...(target && { "data-target": target })}
          {...(collapsable && { "data-toggle": "collapse" })}
          {...(collapsable && subItems && { "data-visible": expanded })}
          {...(collapsable && subItems && { "aria-expanded": expanded })}
          {...spreadProps}
          {...(link && { href: link })}
        >
          <div className="content">
            <div className="flex-fill">{children}</div>
            {collapsable && (
              <i
                className={classNames("fas", expanded ? "fa-chevron-up" : "fa-chevron-down")}
                aria-hidden="true"
              ></i>
            )}
          </div>
        </CustomTag>
        {divider && <BrDivider />}
        {subItems && (
          <BrList hidden={!expanded && collapsable} aria-hidden={!expanded && collapsable}>
            {subItems}
          </BrList>
        )}
      </>
    );
  }
);

export default memo(BrItem);
