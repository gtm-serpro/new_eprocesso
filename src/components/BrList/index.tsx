import { BRList } from "@govbr-ds/core";
import classNames from "classnames";
import React, { memo, useEffect, useRef } from "react";
import useCommonProperties from "../../hooks/useCommonProperties";
import { useMtProps } from "../../hooks/useMtProps";
import { useSpreadProps } from "../../hooks/useSpreadProps";
import useUniqueId from "../../hooks/useUniqueId";
import IMtProps from "../../types/IMtProps";
import BrDivider from "../BrDivider";

export interface ListProps
  extends Omit<React.HTMLAttributes<HTMLDivElement>, "expanded">,
    IMtProps {
  /** Título da lista, opcional. */
  title?: string;
  /** Se a lista é horizontal. */
  horizontal?: boolean;
  /** Se a lista está escondida inicialmente. */
  hidden?: boolean;
  /** Expandida ou não */
  expanded?: boolean;
}

export interface ListRef extends HTMLDivElement {
  element: HTMLDivElement;
}

const BrList = React.forwardRef<ListRef, ListProps>(
  (
    {
      className,
      id,
      children,
      role = "list",
      title,
      horizontal = false,
      hidden = false,
      expanded,
      ...props
    },
    ref
  ) => {
    const fid = useUniqueId(id, "list_____");
    const mtProps = useMtProps(props);
    const spreadProps = useSpreadProps(props);
    const refObject = useRef(null);
    const refDiv = useRef<HTMLDivElement>(null);

    useCommonProperties<ListRef>(ref, refDiv, {
      get element() {
        return refDiv.current;
      },
    });

    useEffect(() => {
      if (refDiv.current && !refObject.current) {
        refObject.current = new BRList("br-list", refDiv.current);
      }
    }, []);

    return (
      <div
        ref={refDiv}
        id={fid}
        className={classNames(
          "br-list",
          horizontal && "horizontal",
          className,

          ...mtProps
        )}
        {...(role && { role: role })}
        {...(hidden && { hidden: "hidden" })}
        {...(expanded && { expanded: "expanded" })}
        {...spreadProps}
      >
        {title && (
          <>
            <div className="header">
              <div className="title">{title}</div>
            </div>
            <BrDivider />
          </>
        )}
        {children}
      </div>
    );
  }
);

export default memo(BrList);
