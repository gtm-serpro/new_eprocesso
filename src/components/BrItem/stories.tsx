import BrCheckbox from "@/components/BrCheckbox";
import { Meta, Story } from "@storybook/react";

import BrItem, { BrItemProps } from ".";

export default {
  title: "Superf√≠cies/Item",
  component: BrItem,
  args: {
    divider: false,
    children: "item x",
  },
} as Meta;

const Template: Story<BrItemProps> = (args) => {
  return <BrItem {...args} />;
};

export const Simples = Template.bind({});

export const ComCheck = Template.bind({});
ComCheck.args = {
  divider: true,
  children: <BrCheckbox label="item x" />,
};
