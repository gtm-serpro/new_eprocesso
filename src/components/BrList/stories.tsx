import BrItem from "@/components/BrItem";
import { Meta, Story } from "@storybook/react";

import BrList, { ListProps } from ".";

const child = (
  <>
    <BrItem>Item 1</BrItem>
    <BrItem>Item 2</BrItem>
  </>
);

export default {
  title: "Superf√≠cies/List",
  component: BrList,
  args: {
    title: "Lista de Itens",
    children: child,
  },
} as Meta;

export const Simples: Story<ListProps> = (args) => <BrList {...args} />;
