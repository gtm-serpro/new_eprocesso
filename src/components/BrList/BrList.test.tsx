import { render, screen } from "@testing-library/react";

import BrItem from "@/components/BrItem";

import BrList from ".";

describe("<BrList />", () => {
  it("deveria renderizar com a classe br-list", () => {
    const { container } = render(<BrList>Conteúdo</BrList>);

    expect(container.firstChild).toHaveClass("br-list");

    expect(container.firstChild).toMatchSnapshot();
  });

  it("deveria renderizar com os itens especificados", () => {
    render(
      <BrList>
        <BrItem>Item 1</BrItem>
        <BrItem>Item 2</BrItem>
      </BrList>
    );

    expect(screen.getByText(/item 1/i)).toBeInTheDocument();
    expect(screen.getByText(/item 2/i)).toBeInTheDocument();
  });

  it("deveria renderizar com a classe horizontal", () => {
    const { container } = render(<BrList horizontal>Conteúdo</BrList>);

    expect(container.firstChild).toHaveClass("horizontal");

    expect(container.firstChild).toMatchSnapshot();
  });
});
