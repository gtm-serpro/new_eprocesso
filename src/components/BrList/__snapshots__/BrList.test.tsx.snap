// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrList /> deveria renderizar com a classe br-list 1`] = `
<div
  class="br-list"
  id="list_____1"
  role="list"
>
  Conteúdo
</div>
`;

exports[`<BrList /> deveria renderizar com a classe horizontal 1`] = `
<div
  class="br-list horizontal"
  id="list_____3"
  role="list"
>
  Conteúdo
</div>
`;
